Migrate from self-hosted to Teleport-enterprise cloud (#40172)
gravitational/teleport: docs/pages/migrating.mdx

- Roles
- Local users
- SSO authentication connectors

Review the [dynamic resources](https://goteleport.com/docs/reference/resources/) list to see if any other resources need to be migrated.  Some common dynamic resources includes:
- windows_desktop
- apps
- dbs
- login_rule

</Details>

<Admonition
  type="tip"
  title="Note"
>

  Trusted Clusters are no longer supported in Teleport Cloud.  You will not be able to migrate trusted cluster resources.

</Admonition>

To achieve this:
- Log in to your existing self-hosted Teleport cluster and export a collection of the above-mentioned dynamic resource configuration using the tctl administrative tool.  An example is shown below:

```code
$ tsh login --proxy=teleport.example.com --user=admin@example.com
$ tctl get roles > roles.yaml 
$ tctl get users > users.yaml 
```

> - Once you have the resource configuration file from the above, proceed to log in to your Teleport Cloud tenant with an admin user and create the resources from the exported files

Reviewer: ```suggestion
- Once you have the resource configuration file from the above, proceed to log in to your Teleport Cloud tenant with an admin user and create the resources from the exported files:

```

```code
$ tsh login --proxy=example.teleport.sh --user=admin@example.com

> $ tctl create -f  roles.yaml
> $ tctl create -f  user.yaml

Reviewer: ```suggestion
$ tctl create -f roles.yaml
$ tctl create -f user.yaml

```

```

For your SSO auth connector, most SSO integrations only work for a single configured endpoint. It is recommended to create a separate SSO connector in your Identity Provider specifically for the Teleport Cloud endpoint, and configure a new Auth Connector in the Teleport Cloud tenant.

> ### Migrate Teleport Agents.

Reviewer: ```suggestion
### Migrate Teleport Agents

```

> To ensure a successful migration process across your infrastructure, start by cataloging the various services you're managing with Teleport, The following resources should be considered for migration:

Reviewer: ```suggestion
To ensure a successful migration process across your infrastructure, start by cataloging the various services you're managing with Teleport. The following resources should be considered for migration:

```

> Teleport Agents (SSH, Kubernetes, Databases, Applications, Desktops)
> Machine ID Bots
> Access Request Plugins

Reviewer: ```suggestion
- Teleport Agents (SSH, Kubernetes, Databases, Applications, Desktops)
- Machine ID Bots
- Access Request Plugins

```

The following procedure is required to successfully migrate your Teleport agents:

> - Log in to your Teleport Enterpise cloud Tenant, You'll need to authenticate to access your Teleport Cloud resources and management interfaces.

Reviewer: ```suggestion
- Log in to your Teleport Enterpise Cloud tenant. You'll need to authenticate to access your Teleport Cloud resources and management interfaces.

```

- Obtain a valid join token: We recommend using a [Delegated Join Method](https://goteleport.com/docs/reference/join-methods/#delegated-join-methods)

> If using ephemeral tokens, ensure you specify the appropriate token type to match the Teleport services. Token types can include `node`, `app`, `Kube`, `DB`, `WindowsDesktop`, and others depending on the service you wish to join to your Teleport cluster.

Reviewer: I wasn't sure whether you were mentioning the `tctl tokens add` syntax here or the delegated join `Node`, `App`, `DB` type syntax so I went with the former.

```suggestion
If using ephemeral tokens, ensure you specify the appropriate token type to match the Teleport services. Token types can include `node`, `app`, `kube`, `db`, `windowsdesktop` and others depending on the service you wish to join to your Teleport cluster.

```

In the following example, a new token is created with the TTL of fifteen minutes:

```code
$ tctl tokens add --type node,app,db --ttl 15m
```

> In this command, we assigned the token the `node`, `app` and `db` type, indicating that it will belong to an SSH Service, Database Service and Application Service instance.

Reviewer: ```suggestion
In this command, we assigned the token the `node`, `app` and `db` type, indicating that it will allow an agent to join which is running the Teleport `ssh_service`, `db_service` and `app_service`.

```

Copy the token so you can use it later in this guide.

- Stop the Teleport service on the agent (if applicable)

> - Update the `proxy_server` or `auth_server` field in the agent configuration file to point to the address of your Teleport cloud proxy. By default, the configuration is located in the `/etc/teleport.yaml` directory

Reviewer: ```suggestion
- Update the `proxy_server` or `auth_servers` field in the agent configuration file to point to the address of your Teleport Cloud cluster. By default, the configuration is located in the `/etc/teleport.yaml` directory.

```

> ```code
> proxy_server:
> - example.teleport.sh:443
> ```

Reviewer: We should try and teach people to move to newer config versions at the same time IMO, which means using `version: v3` and specifying `proxy_server` instead of `auth_servers`.

````suggestion
```yaml
version: v3
teleport:
  proxy_server: example.teleport.sh:443
```
````

> - Update the auth_token field with the newly generated token. 

Reviewer: ````suggestion
- Update the `auth_token` or `join_params.token_name` field with the newly generated token.

```yaml
teleport:
  join_params:
    method: token
    token_name: new-token-goes-here
```

````

> - Delete the local agent cache to force the agent to rejoin the new Teleport Cloud cluster. By default, the data is located in the  `/var/lib/teleport` directory

Reviewer: ````suggestion
- Delete the local agent cache to force the agent to rejoin the new Teleport Cloud cluster. By default, the data is located in the  `/var/lib/teleport` directory:

```code
rm -rf /var/lib/teleport
```

````

> - Start the Teleport process on each agent or recycle the pod to apply the new configuration. This will cause the agents to re-register with the Teleport cloud cluster and obtain new certificates signed by the cloud cluster's certificate authority.

Reviewer: ```suggestion
- Start the Teleport process on each agent or recycle the pod to apply the new configuration. This will cause the agents to re-register with the Teleport Cloud cluster and obtain new certificates signed by the Teleport Cloud cluster's certificate authority.

```

As new services come online, they start sending heartbeat requests every few seconds to the Auth Service. This allows users to explore cluster membership. 

> Note: ​​Migration of agents can be done all at once or gradually, depending on your business requirements. If running teleport at scale, you'll generally want to leverage the use of a configuration management tool to automate and streamline the process of carrying out the actions involved in migrating agent configurations. 

Reviewer: ```suggestion
Note: ​​Migration of agents can be done all at once or gradually, depending on your business requirements. If running Teleport at scale, you'll generally want to leverage the use of a configuration management tool to automate and streamline the process of carrying out the actions involved in migrating agent configurations. 

```

> ### Verify end User Access and performance.

Reviewer: ```suggestion
### Verify end User Access and performance

```

> Once the migration is complete, validate that the resources are present in the Teleport cluster and verify their status and connectivity to ensure they are properly registered and available for connections. You can either check via the Web UI or use tctl to get a list of all resources and verify their registration and status. 

Reviewer: ```suggestion
Once the migration is complete, validate that the resources are present in the Teleport cluster and verify their status and connectivity to ensure they are properly registered and available for connections. You can either check via the Web UI, or use `tctl` to get a list of all resources and verify their registration and status. 

```

For example, to list all nodes registered with the Teleport cluster, you can run the command below

```
tctl nodes ls
```

Similarly, to list all registered Kubernetes clusters, you can run the command below:

```

> tctl get kube_services

Reviewer: `tctl kube ls` is probably a better bet as it will show the name, version and labels of each in a more human-readable form.

```suggestion
tctl kube ls
```

Reviewer: Can we also add sections for `tctl db ls`, `tctl apps ls` and `tctl get windows_desktop` here so people have the full range in one place?

```

> These tctl commands will show you the resources registered within your Teleport cluster, and you can verify their availability for connections

Reviewer: ```suggestion
These `tctl` commands will show you the resources registered within your Teleport cluster, and you can verify their availability for connections.

```

- Ensure that end users have the expected SSO access to your infrastructure and workflow.

> Additionally, establish break glass access procedures to ensure access to infrastructure in case of Teleport Cloud unavailability.

Reviewer: (how should people do this?!)

### Conclusion

> Migrating from a self-hosted Teleport deployment to Teleport Cloud offers scalability, reliability, and ease of management. By following the steps outlined in this guide, you can transition your Teleport self-hosted deployment to the cloud successfully while ensuring minimal disruption to your operations.

Reviewer: ```suggestion
Migrating from a self-hosted Teleport deployment to Teleport Cloud offers scalability, reliability, and ease of management. By following the steps outlined in this guide, you can transition your Teleport self-hosted deployment to Teleport Cloud successfully while ensuring minimal disruption to your operations.

```
